src/features/defi/providers/osmosis/components/OsmosisManager/Deposit/components/Deposit.tsx

```diff
@@ -0,0 +1,375 @@
+    toast,
+    translate,
+  ])
+
+  /** Returns an array containing the allocation fraction and shareOutAmount,
+   * calculated from the form input parameters */
+  /** Share out amount is the exact number of tokens expected to be received. See documentation below:
+   * https://docs.osmosis.zone/osmosis-core/modules/gamm/#join-pool
+   * https://github.com/osmosis-labs/osmosis-frontend/blob/d3badcee48f54e61e6a6117c294c0725822a4100/packages/stores/src/ui-config/manage-liquidity/add-liquidity.ts#L371
+   * NOTE: This calculation may need to be modified to support pools with unequal weights at a later date. The Osmosis frontend does not currently do this.
+   */
+  const calculateAllocations = useCallback(
+    (
+      inputAsset: Asset,
+      inputAssetAmount: string,
+    ): { allocationFraction: string; shareOutAmount: string } => {
+      const ret = { allocationFraction: '0', shareOutAmount: '0' }
+      if (!state || !state.poolData) {
+        return ret
+      }
+      const poolAssets = state?.poolData?.pool_assets
+
+      let { assetReference: inputAssetReference } = fromAssetId(inputAsset.assetId)
+      if (inputAssetReference === ASSET_REFERENCE.Osmosis) inputAssetReference = 'uosmo'
+
+      if (!poolAssets) return ret
+      const poolAssetIndex = poolAssets.findIndex(asset =>
+        asset.token.denom.toLowerCase().includes(inputAssetReference.toLowerCase()),
+      )
+      if (poolAssetIndex < 0) return ret
+
+      const inputAssetAmountBaseUnits = bnOrZero(inputAssetAmount)
+        .pow(10, bnOrZero(inputAsset.precision))
+        .toString()
+
+      const poolTotalSharesBaseUnits = state?.poolData?.total_shares?.amount
+      if (!poolTotalSharesBaseUnits) {
+        moduleLogger.error(`Could not get total shares for pool ${state?.poolData?.id}`)
+        return ret
+      }
+
+      const poolAssetAmountBaseUnits = poolAssets[poolAssetIndex]?.token.amount
+
+      if (bnOrZero(poolAssetAmountBaseUnits).eq(bn(0))) return ret
+
+      /** allocation_fraction = input_asset_amount / (pool_asset_amount + pool_asset_amount)
+       * This represents the fraction of the pool that the user will own after once the liquiditt
+       * has been provided
+       */
+      const allocationFraction = bnOrZero(inputAssetAmountBaseUnits)
+        .div(bnOrZero(poolAssetAmountBaseUnits).plus(inputAssetAmountBaseUnits))
+        .toString()
+
+      /** share_out_amount = total_pool_shares * (input_asset_amount / pool_asset_amount)
+       * This represents the number of pool shares that the user will receive in exchange
+       * for the provided liquidity
+       */
+      const shareOutAmountBaseUnits = bnOrZero(poolTotalSharesBaseUnits)
+        .multipliedBy(
+          bnOrZero(inputAssetAmountBaseUnits).dividedBy(bnOrZero(poolAssetAmountBaseUnits)),
+        )
+        .toFixed(0)
+        .toString()
+
+      ret.allocationFraction = allocationFraction
+      ret.shareOutAmount = shareOutAmountBaseUnits
+      return ret
```

Comment:
Object mutation could be avoided by moving the whole logic into the if at the beginning of the callback and returning a fresh object here, and the initial object outside of the if statement.

Alternatively, if the intent is to handle loading states in the render cycles when `state.poolData` isn't yet defined, the signature of this could be
`{ allocationFraction: string; shareOutAmount: string } | undefined` which would allow to check for `!allocation` and don't fire actions while we're loading